### PR TITLE
fix(consumer): Immediatly stop polling, abort current request

### PR DESCRIPTION
**Pull request by yehor-manzhula**
## Description
Stopping polling timeout and current polling request.

## Motivation and Context
Stop method doesn't gracefully terminate polling:
1. Stop method doesn't abort current polling request
2. "Stopped" event emitted after polling timeout execution, that doesn't make any sense.

Actuall polling stop depends on: 
 waitTimeSeconds + Messages processing time + pollingWaitTimeMs 
[Multiple polls invoked](https://github.com/bbc/sqs-consumer/issues/234)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Review by ankitgehlot123 — APPROVED** [review]

File: src/consumer.ts [hunk 1]
```diff
@@ -68,6 +68,14 @@ function isConnectionError(err: Error): boolean {
   return false;
 }
 
+function isAbortedError(err: Error): boolean {
+  if (err instanceof SQSError) {
+    return err.statusCode === 200 || err.code === 'RequestAbortedError';
```
Comment (nicholasgriffintn) [inline src/consumer.ts:73]: Not sure we want to consider everything with a status code 200 as an abort error...

File: src/consumer.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { AWSError, SQS } from 'aws-sdk';
+import { AWSError, SQS, Request as AWSRequest } from 'aws-sdk';
```
Comment (nicholasgriffintn) [inline src/consumer.ts:1]: This will all need to be updated to AWS SDK V3

File: src/consumer.ts [hunk 3]
```diff
@@ -253,7 +283,14 @@ export class Consumer extends EventEmitter {
...
       return await this.sqs.receiveMessage(params).promise();
     } catch (err) {
-      throw toSQSError(err, `SQS receive message failed: ${err.message}`);
+      const sqsError = toSQSError(
+        err,
+        `SQS receive message failed: ${err.message}`
+      );
+
+      if (!isAbortedError(sqsError)) {
+        throw sqsError;
```
Comment (nicholasgriffintn) [inline src/consumer.ts:292]: Do we not want to throw this error still? I think we probably do...

File: src/consumer.ts [hunk 4]
```diff
@@ -333,7 +370,6 @@ export class Consumer extends EventEmitter {
 
   private poll(): void {
     if (this.stopped) {
-      this.emit('stopped');
```
Comment (nicholasgriffintn) [inline src/consumer.ts:336]: I'm unsure why this has been removed.

File: src/consumer.ts [hunk 5]
```diff
@@ -133,6 +141,12 @@ export class Consumer extends EventEmitter {
   private sqs: SQS;
   private shouldDeleteMessages: boolean;
 
+  private currentPollingTimeoutId: NodeJS.Timeout | null;
+  private currentPollingRequest: AWSRequest<
```
Comment (nicholasgriffintn) [inline src/consumer.ts:145]: Not to sure about this overall implementation.

**Review by nicholasgriffintn — CHANGES_REQUESTED** [review]

**Comment by nicholasgriffintn** [comment]
Closing this PR given the age and changes requested above, if you would still like to contribute please open a new PR, we may look into this separately as well.

**Pull request closed without merging**